packages/core/src/hydration/i18n.ts

```diff
@@ -0,0 +1,325 @@
+
+  // Claim the current node and move to the next one.
+  CLAIM,
+}
+
+/**
+ * Marks the current RNode as the hydration root for the given
+ * AST node.
+ */
+function markHydrationRoot(
+    context: I18nHydrationContext, state: I18nHydrationState, astNode: I18nNode,
+    options: MarkOptions) {
+  const noOffsetIndex = astNode.index - HEADER_OFFSET;
+  const {disconnectedNodes} = context;
+
+  if (state.isConnected) {
+    const currentNode = state.currentNode;
+    context.i18nMap.set(noOffsetIndex, currentNode);
+
+    if (currentNode) {
+      if (options === MarkOptions.CLAIM) {
+        state.currentNode = currentNode.nextSibling;
+      }
+    }
+
+    // We expect the node to be connected, so ensure that it
+    // is not in the set, regardless of whether we found it,
+    // so that the downstream error handling can provide the
+    // proper context.
+    disconnectedNodes.delete(noOffsetIndex);
+  } else {
+    disconnectedNodes.add(noOffsetIndex);
+  }
+}
+
+/**
+ * Skip over some nodes during hydration.
+ *
+ * Note: we use this instead of `siblingAfter` as it's expected that
+ * sometimes we might encounter null nodes. In those cases, we want to
+ * defer to downstream error handling to provide proper context.
+ */
+function skipHydrationNodes(state: I18nHydrationState, skip: number) {
+  let currentNode = state.currentNode;
+  for (let i = 0; i < skip; i++) {
+    if (!currentNode) {
+      break;
+    }
+    currentNode = currentNode?.nextSibling ?? null;
+  }
+  state.currentNode = currentNode;
+}
+
+/**
+ * Fork the given state into a new state for hydrating children.
+ */
+function forkChildHydrationState(state: I18nHydrationState) {
+  const currentNode = state.currentNode as Node | null;
+  return {currentNode: currentNode?.firstChild ?? null, isConnected: state.isConnected};
+}
+
+function prepareI18nBlockForHydrationImpl(lView: LView, index: number) {
+  if (!isI18nHydrationSupportEnabled()) {
+    return;
+  }
+
+  const hydrationInfo = lView[HYDRATION];
+  if (!hydrationInfo) {
+    return;
+  }
+
+  const tView = lView[TVIEW];
+  const tI18n = tView.data[index] as TI18n;
+  ngDevMode &&
+      assertDefined(
+          tI18n, 'Expected i18n data to be present in a given TView slot during hydration');
+
+  const firstAstNode = tI18n.ast[0];
+  if (firstAstNode) {
+    // We begin hydration at the RNode for the first child. Top-level AST nodes will always
+    // have a valid TNode, so we can use the normal `locateNextRNode` to find it. It's OK if
+    // nothing is located, as that also means that there is nothing to clean up. Downstream
+    // error handling will detect this and provide proper context.
+    const tNode = tView.data[firstAstNode.index] as TNode;
+    const rootNode = locateNextRNode(hydrationInfo, tView, lView, tNode) as Node | null;
+
+    const disconnectedNodes = initDisconnectedNodes(hydrationInfo) ?? new Set();
+    const i18nMap = hydrationInfo.i18nNodes ??= new Map<number, RNode|null>();
```

Comment:
I'd like to propose that we replace `i18nMap` with `i18nNodes`, so it's easier to make a "connection" to the `i18nNodes` field in `hydrationInfo` object.
```suggestion
    const i18nNodes = hydrationInfo.i18nNodes ??= new Map<number, RNode|null>();
```

Reply:
Ah, good catch, my mistake.